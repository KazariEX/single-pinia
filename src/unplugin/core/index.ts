import { camelize, capitalize } from "@vue/shared";
import { babelParse, getLang, walkAST } from "ast-kit";
import { generateTransform, MagicStringAST } from "magic-string-ast";
import type t from "@babel/types";

export function transformSinglePinia(code: string, id: string) {
    const lang = getLang(id);
    const program = babelParse(code, lang);

    const s = new MagicStringAST(code);
    const exportNames: string[] = [];

    let defined = false;

    for (const node of program.body) {
        if (
            node.type === "ExpressionStatement"
            && node.expression.type === "CallExpression"
            && node.expression.callee.type === "Identifier"
            && node.expression.callee.name === "defineStore"
        ) {
            defined = true;
            const [arg0] = node.expression.arguments;

            let id = "";
            let arg0End = 0;
            if (arg0?.type === "StringLiteral") {
                id = arg0.value;
                arg0End = arg0.end!;
            }

            s.appendLeft(node.start!, `export const use${
                capitalize(camelize(id))
            }Store = `);

            s.prependLeft(arg0End, `, (helper) => {`);

            s.move(node.end!, code.length, arg0End);
        }
        else if (
            defined
            && node.type === "ExportNamedDeclaration"
            && node.declaration
        ) {
            const decl = node.declaration;
            if (decl.type === "VariableDeclaration") {
                const decls = decl.declarations;
                for (const decl of decls) {
                    exportNames.push(...collectVars(decl.id, []));
                }
            }
            else if (decl.type === "FunctionDeclaration") {
                exportNames.push(decl.id!.name);
            }
            s.remove(node.start!, decl.start!);
        }
    }

    s.appendLeft(code.length, "\nreturn { ");
    for (const name of exportNames) {
        s.appendLeft(code.length, `${name}, `);
    }
    s.appendLeft(code.length, "};\n}");

    return generateTransform(s, id);
}

function collectVars(
    id: t.Node,
    result: string[]
) {
    if (id.type === "Identifier") {
        result.push(id.name);
    }
    else if (id.type === "ObjectPattern") {
        for (const prop of id.properties) {
            if (prop.type === "RestElement") {
                collectVars(prop.argument, result);
            }
            else {
                collectVars(prop.value, result);
            }
        }
    }
    else if (id.type === "ArrayPattern") {
        for (const elem of id.elements) {
            if (elem) {
                collectVars(elem, result);
            }
        }
    }
    else {
        walkAST(id, {
            enter(node) {
                collectVars(node, result);
            }
        });
    }
    return result;
}