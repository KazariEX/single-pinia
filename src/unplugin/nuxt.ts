import { addVitePlugin, addWebpackPlugin, createIsIgnored, createResolver, defineNuxtModule } from "@nuxt/kit";
import { camelize, capitalize } from "@vue/shared";
import { basename } from "pathe";
import { type Import, normalizeScanDirs, type ScanDirExportsOptions, scanFilesFromDir } from "unimport";
import vite from "./vite";
import webpack from "./webpack";
import type { Options } from "../types";

export interface ModuleOptions extends Options {}

export default defineNuxtModule<ModuleOptions>({
    meta: {
        name: "nuxt-single-pinia",
        configKey: "singlePinia"
    },
    setup(options, nuxt) {
        const { resolve } = createResolver(import.meta.url);

        addVitePlugin(() => vite(options));
        addWebpackPlugin(() => webpack(options));

        let storesDirs = [resolve(nuxt.options.srcDir, "stores")];
        if ("pinia" in nuxt.options) {
            storesDirs = (
                ((nuxt.options.pinia ?? {}) as import("@pinia/nuxt").ModuleOptions).storesDirs ??= storesDirs
            ).map((dir) => resolve(nuxt.options.rootDir, dir));
        }

        const isIgnored = createIsIgnored(nuxt);
        nuxt.hook("imports:extend", async (imports) => {
            const options: ScanDirExportsOptions = {
                fileFilter: (file) => !isIgnored(file)
            };
            const normalizedDirs = normalizeScanDirs(storesDirs, options);
            const files = await scanFilesFromDir(normalizedDirs, options);

            for (let i = 0; i < imports.length; i++) {
                const item = imports[i];
                if (files.includes(item.from)) {
                    imports.splice(i, 1);
                    i--;
                }
            }

            const scannedImports = files.map<Import>((file) => {
                const name = basename(file).split(".")[0];
                return {
                    from: file,
                    name: `use${capitalize(camelize(name))}Store`
                };
            });

            imports.push(...scannedImports);
        });
    }
});