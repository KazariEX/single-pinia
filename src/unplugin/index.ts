import { createUnplugin } from "unplugin";
import { defaultPattern } from "../shared";
import { transformSetupPinia } from "./core";
import type { Options } from "../types";

const unplugin = createUnplugin<Options>((options) => ({
    name: "setup-pinia",
    enforce: "pre",
    transformInclude(id) {
        const {
            pattern = defaultPattern
        } = options ?? {};
        return pattern.test(id);
    },
    transform: transformSetupPinia
}));

export default unplugin;