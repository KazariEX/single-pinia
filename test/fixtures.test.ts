import { testFixtures } from "@sxzz/test-utils";
import { describe } from "vitest";
import { transformSinglePinia } from "../src/unplugin/core";

describe("transform", async () => {
    await testFixtures(
        import.meta.glob<string>("./fixtures/*.ts", {
            eager: true,
            query: "?raw",
            import: "default"
        }),
        (_, id, code) => transformSinglePinia(code, id)?.code
    );
});