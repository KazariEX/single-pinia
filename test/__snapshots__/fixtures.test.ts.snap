// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform > ./fixtures/counter.ts 1`] = `
"import { defineStore } from "pinia";
import { ref } from "vue";

export const useCounterStore = defineStore("counter", (helper) => {

const count = ref(0);

function increment() {
    count.value++;
}

function decrement() {
    count.value--;
}
return { count, increment, decrement, };
}, {
    persist: true
});"
`;
